creator-node/src/snapbackSM/snapbackSM.js

```diff
@@ -347,67 +422,196 @@ class SnapbackSM {
+   * the value of `this.highestReconfigModeEnabled` and the type of reconfig that needs to be issued. See `RECONFIG_MODE_KEYS` variable
+   * for more information.
+   *
    * @param {Object} param
    * @param {string} param.primary current user's primary endpoint
    * @param {string} param.secondary1 current user's first secondary endpoint
    * @param {string} param.secondary2 current user's second secondary endpoint
    * @param {string} param.wallet current user's wallet address
-   * @param {string[]} param.unhealthyReplicasSet an array of endpoints of unhealthy replica set nodes
+   * @param {Set<string>} param.unhealthyReplicasSet a set of endpoints of unhealthy replica set nodes
    * @param {string[]} param.healthyNodes array of healthy Content Node endpoints used for selecting new replica set
    * @returns {Object}
    * {
-   *  newPrimary: {string} the endpoint of the newly selected primary,
-   *  newSecondary1: {string} the endpoint of the newly selected secondary #1,
-   *  newSecondary2: {string} the endpoint of the newly selected secondary #2
+   *  newPrimary: {string | null} the endpoint of the newly selected primary or null,
+   *  newSecondary1: {string | null} the endpoint of the newly selected secondary #1,
+   *  newSecondary2: {string | null} the endpoint of the newly selected secondary #2,
+   *  issueReconfig: {boolean} flag to issue reconfig or not
    * }
    */
   async determineNewReplicaSet ({ primary, secondary1, secondary2, wallet, unhealthyReplicasSet, healthyNodes }) {
-    let currentNodeHasExistingUserState = true
-    let i = 0
-    let newSecondary
-    while (currentNodeHasExistingUserState && i++ < MAX_SELECT_NEW_REPLICA_SET_ATTEMPTS) {
-      // Make sure current replica set will not be chosen as potential nodes for the new replica set
-      const healthyPotentialPeers = healthyNodes.filter(node => node !== primary && node !== secondary1 && node !== secondary2)
+    let response = { newPrimary: null, newSecondary1: null, newSecondary2: null, issueReconfig: false }
+
+    // If entire replica set is unhealthy, select an entire new replica set
+    if (unhealthyReplicasSet.size === NUMBER_OF_REPLICA_SET_NODES) {
+      this.logWarn(`[determineNewReplicaSet] Entire replica set=[${Array.from(unhealthyReplicasSet)}] is unhealthy. Not issuing new replica set.`)
+      response.issueReconfig = false
+      return response
+    }
+
+    const newReplicaNodes = await this.selectRandomReplicaSetNodes({
+      unhealthyReplicasSet,
+      healthyNodes,
+      wallet
+    })
+
+    // Perform last ditch effort to see if primary is truly unhealthy. If the primary responds
+    // with a 200 from the health check, remove from `unhealthyReplicaSet`. Else, continue with
+    // the new replica set determination logic.
+    const secondariesToClockMap = await this.fetchClockValues(secondary1, secondary2, wallet)
+    unhealthyReplicasSet = await this.finalizeUnhealthyReplicaSet({
```

Comment:
i'm pretty confused between this and fetchClockValues. not really sure how both are valuable. also i think the rationale with multiple checks is we don't have false negatives, but i think if our peerSetManager health checks are configured correctly seems redundant?

Reply:
> i'm pretty confused between this and fetchClockValues. not really sure how both are valuable.

in the scenario where a primary is unhealthy, we want to issue a reconfig with either one of the two secondaries with the greater clock value getting chosen as the new primary. in the case that the the clock fetches return expected secondary clock values, `finalizeUnhealthyReplicaSet` will be a no-op for handling secondaries. 

however, if when trying to fetch clock values for secondaries fails after `n` attempts (e.g. the map has a `null` clock value in the returned map), we can probably conclude that the secondary is actually unhealthy. with this assumption, we want to update the unhealthy replica set by adding the relevant secondaries.

> also i think the rationale with multiple checks is we don't have false negatives, but i think if our peerSetManager health checks are configured correctly seems redundant?

yes and no on the redundancy.
_no_ -- if primaries are in the unhealthy set, we want to be extra sure that it is indeed unhealthy as the primary may have data that the secondaries do not have. this is why in this updated PR, if we see a primary as unhealthy:
1. do a last minute attempt to see if the health check returns a 200. 
2. if the primary responds with 200, remove from unhealthy set. else, mark as visited for that user.
3. wait moduloBase iterations
4. if the primary is unhealthy again and has been visited before, issue the reconfig

these above steps are an extra layer of security to ensure that if a reconfig is to be issued that it was truly necessary

_yes_ -- as described earlier, there are already rigorous health checks in place for nodes in the peer set. checking clock values as an extra layer of confidence that a node is un/healthy is unnecessary. however, the main point of this was to handle selecting a proper secondary as the primary if need be. also marking a secondary as healthy or not was more so an after thought

Reply:
I guess my question mostly comes down to what the drawback is if the reconfig happens even if a secondary was down and is up. Taking it a step further, even simplifying to one health check. 

- [ ] Also I guess another concern is this adds a lot of complexity and splinters a lot of the health logic where there's so many branches it's hard to keep track. Either we should have a good reason for having this and we should properly abstract it or maybe it's worth simplifying?

Reply:
this is a fair point. I think this reconfig state machine logic has gotten more complex with these additional health checks. I am in favor of simplifying the logic -- we just need to be aware of the pros/cons that come with a simplification. 

let's chat tomorrow on this on video with @SidSethi and @hareeshnagaraj

Reply:
Yep sounds good!